tests/api/test_share.py

```diff
@@ -330,139 +509,393 @@ def test_list_dataset_share_objects(
+                    count
+                    page
+                    pages
+                    hasNext
+                    hasPrevious
+                    nodes {
+                        itemUri
+                        shareItemUri
+                        itemType
+                        itemName
+                        status
+                        action
+                    }
                 }
             }
         }
-    }
     """
+
     response = client.query(
-        get_share_object_query,
-        username=dataset1.owner,
+        getShareObjectQuery,
+        username=user.userName,
+        shareUri=share1.shareUri,
         groups=[group.name],
-        datasetUri=dataset1.datasetUri,
     )
-    share_object = response.data.getDataset.shares.nodes[0]
-    this_dataset_tables = response.data.getDataset.tables.nodes
-    print(this_dataset_tables)
-    query = """
+
+    print('Response from getShareObject: ', response)
+
+    assert (
+        response.data.getShareObject.status
+        == dataall.api.constants.ShareObjectStatus.Draft.name
+    )
+
+    # Given existing shareable items (-> fixture)
+    shareableItem = response.data.getShareObject.get('items').nodes[0]
+    itemUri = shareableItem['itemUri']
+    itemType = shareableItem['itemType']
+    shareItemUri = shareableItem['shareItemUri']
+    assert shareItemUri is None
+
+    # When
+    addSharedItemQuery = """
     mutation AddSharedItem($shareUri:String!,$input:AddSharedItemInput){
         addSharedItem(shareUri:$shareUri,input:$input){
             shareUri
             shareItemUri
             itemUri
+            status
+            action
         }
     }
     """
 
     response = client.query(
-        query,
-        username=dataset1.owner,
-        shareUri=share_object.shareUri,
+        addSharedItemQuery,
+        username=user.userName,
+        groups=[group.name],
+        shareUri=share1.shareUri,
         input={
-            'itemUri': 'foo',
-            'itemType': dataall.api.constants.ShareableType.Table.name,
+            'itemUri': itemUri,
+            'itemType': itemType,
         },
-        groups=[group.name],
     )
 
-    assert 'ResourceNotFound' in response.errors[0].message
+    print('Response from addSharedItem: ', response)
 
-    response = client.query(
-        query,
-        username='noneofmybusiness',
-        groups=['anonymous'],
-        shareUri=share_object.shareUri,
-        input={
-            'itemUri': 'foo',
-            'itemType': dataall.api.constants.ShareableType.Table.name,
-        },
-    )
-    assert 'UnauthorizedOperation' in response.errors[0].message
+    # Then shared item was added to share object in status PendingApproval
+    assert response.data.addSharedItem.shareUri == share1.shareUri
+    assert response.data.addSharedItem.status == \
+        dataall.api.constants.ShareItemStatus.PendingApproval.name
+
+
+@pytest.mark.dependency(depends=["test_add_share_item"])
```

Comment:
Regarding these test for the share api we refactored our forked version again into a setup where each test is independent of one another and instead the required state is constructed in each test with a cleanup after each tests which ensures that each test can run independent from another.

For this we first moved all the graphql queries/mutations into separate functions, so the calls can be reused in multiple test cases:

    def add_share_item(shareUri, itemUri, itemType, client, userName, groupUri):
    query = """
    mutation AddSharedItem($shareUri:String!,$input:AddSharedItemInput){
        addSharedItem(shareUri:$shareUri,input:$input){
            shareUri
            shareItemUri
            itemUri
            status
            action
        }
    }
    """

    response = client.query(
        query,
        username=userName,
        groups=[groupUri],
        shareUri=shareUri,
        input={
            "itemUri": itemUri,
            "itemType": itemType,
        },
    )

    print("Response from addSharedItem: ", response)

    return response


We added a cleanup part to the share fixture:

     @pytest.fixture(scope="function")
     def share1(env2, user, group, user2, group2, dataset1, share, client, db):
     share1 = share(
         environmentUri=env2.environmentUri, 
         userName=user2.userName, 
         groupUri=group2.groupUri, 
         datasetUri=dataset1.datasetUri
     )

    yield share1

    # Cleanup share
    # First try to direct delete
    delete_share_object_response = delete_share_object(shareUri=share1.shareUri, client=client, 
    userName=user2.userName, groupUri=group2.groupUri)

    if delete_share_object_response.data.deleteShareObject == True:
        return

    # Then delete via revoke if still items present
    revoke_all(share1.shareUri, client, user2.userName, group2.groupUri)

    succesfull_processing_of_share_object(db, share1.shareUri)

    delete_share_object(
        shareUri=share1.shareUri,
        client=client, 
        userName=user2.userName, 
        groupUri=group2.groupUri
    )

Finally the test cases are setting up there own required test state:

     def test_submit_share_request_with_share_item(share1, table1, client, user2, group2):
     # Given
     # Existing share object in status Draft (-> fixture share1)
     # and share item added
     add_share_item(share1.shareUri, table1.tableUri, dataall.api.constants.ShareableType.Table.name, client, 
     user2.userName, group2.groupUri)

     get_share_object_response = get_share_object(
        share1.shareUri,
        client,
        user2.userName,
        group2.groupUri,
    )

    assert get_share_object_response.data.getShareObject.status == 
    dataall.api.constants.ShareObjectStatus.Draft.name

    shareItem = get_share_object_response.data.getShareObject.get("items").nodes[0]
    assert shareItem.status == dataall.api.constants.ShareItemStatus.PendingApproval.value
    assert get_share_object_response.data.getShareObject.get("items").count == 1

    # When
    # Submit share object
    submit_share_object_response = submit_share_object(
        share1.shareUri,
        client,
        user2.userName,
        group2.groupUri,
    )
    # Then ....

Reply:
Hi @degoldner thank you for the suggestion! Would you mind open a PR to this branch for this? Then I will have a look at the changes. I will continue with the remaining implementation tasks today/tomorrow, let's see if we can have this done by this week or early next week